// carbon-components-angular/fesm2015/carbon-components-angular-skeleton.js
/**
 *
 * carbon-angular v0.0.0 | carbon-components-angular-skeleton.js
 *
 * Copyright 2014, 2021 IBM
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0

 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */


import { Component, Input, NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';

class SkeletonPlaceholder {
}
SkeletonPlaceholder.decorators = [
    { type: Component, args: [{
                selector: "ibm-skeleton-placeholder",
                template: `
		<div class="bx--skeleton__placeholder"></div>`
            }] }
];

class SkeletonText {
    constructor() {
        this.lines = 5;
        this.minLineWidth = 100;
        this.maxLineWidth = 300;
    }
    /**
     * Returns a random width in pixels based off a min width, and a max width.
     */
    getRandomInt(min, max) {
        return `${Math.floor(Math.random() * (max - min + 1) + min) + "px"}`;
    }
    ngOnChanges() {
        // Creates an array of length defined by input lines with content from
        // 0 to lines - 1, maps each value to a random width in pixels.
        this.lineWidths =
            Array.from(Array(this.lines).keys())
                .map(num => this.getRandomInt(this.minLineWidth, this.maxLineWidth));
    }
}
SkeletonText.decorators = [
    { type: Component, args: [{
                selector: "ibm-skeleton-text",
                template: `
		<p
			*ngFor="let width of lineWidths"
			class="bx--skeleton__text"
			[style.width]="width">
		</p>
	`
            }] }
];
SkeletonText.propDecorators = {
    lines: [{ type: Input }],
    minLineWidth: [{ type: Input }],
    maxLineWidth: [{ type: Input }]
};

class SkeletonModule {
}
SkeletonModule.decorators = [
    { type: NgModule, args: [{
                declarations: [
                    SkeletonPlaceholder,
                    SkeletonText
                ],
                exports: [
                    SkeletonPlaceholder,
                    SkeletonText
                ],
                imports: [
                    CommonModule
                ]
            },] }
];

/**
 * Generated bundle index. Do not edit.
 */

export { SkeletonPlaceholder, SkeletonText, SkeletonModule };
