// carbon-components-angular/esm2015/dialog/tooltip/ellipsis-tooltip.directive.js
/**
 *
 * carbon-angular v0.0.0 | ellipsis-tooltip.directive.js
 *
 * Copyright 2014, 2021 IBM
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0

 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */


import { Directive, HostListener } from "@angular/core";
import { TooltipDirective } from "./tooltip.directive";
import { DialogService } from "../dialog.service";
/**
 * A directive that creates a tooltip `Dialog` for exposing truncated text.
 *
 * class: EllipsisTooltip (extends PopoverDirective)
 *
 * selector: `nEllipsisTooltip`
 *
 * ```html
 * <div class="ellipsis" nEllipsisTooltip>Tooltip for ellipsis because I can and I am really really long</div>
 * ```
 */
export class EllipsisTooltip extends TooltipDirective {
    /**
     * If text is truncated, this appends the text to the dialog as content.
     * @returns null
     * @memberof EllipsisTooltip
     */
    updateTooltipContent() {
        if (this.elementRef.nativeElement.scrollWidth <= this.elementRef.nativeElement.offsetWidth) {
            this.disabled = true;
            return;
        }
        this.disabled = false;
        this.dialogConfig.content = this.elementRef.nativeElement.innerText;
    }
    onClick() {
        if (this.trigger === "click") {
            this.updateTooltipContent();
        }
    }
    onHover() {
        if (this.trigger === "hover" || this.trigger === "mouseenter") {
            this.updateTooltipContent();
        }
    }
    onFocus() {
        this.updateTooltipContent();
    }
}
EllipsisTooltip.decorators = [
    { type: Directive, args: [{
                selector: "[ibmEllipsisTooltip]",
                providers: [
                    DialogService
                ]
            },] }
];
EllipsisTooltip.propDecorators = {
    onClick: [{ type: HostListener, args: ["click",] }],
    onHover: [{ type: HostListener, args: ["mouseenter",] }],
    onFocus: [{ type: HostListener, args: ["focus",] }]
};
