// carbon-components-angular/esm2015/dialog/tooltip/tooltip-definition.component.js
/**
 *
 * carbon-angular v0.0.0 | tooltip-definition.component.js
 *
 * Copyright 2014, 2021 IBM
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0

 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */


import { Component, Input, HostBinding } from "@angular/core";
export class TooltipDefinition {
    constructor() {
        this.id = `definition-tooltip-${TooltipDefinition.tooltipItemCount++}`;
        /**
         * The placement in which the `TooltipDefinition` appears.
         * Set to `"top"` to have it positioned above the trigger text
         */
        this.placement = "bottom";
        this.alignment = "start";
        this.className = true;
        this.accessibility = true;
    }
}
TooltipDefinition.tooltipItemCount = 0;
TooltipDefinition.decorators = [
    { type: Component, args: [{
                selector: "ibm-tooltip-definition",
                template: `
		<button
			type="button"
			[ngClass]="{
				'bx--tooltip--bottom' : placement === 'bottom',
				'bx--tooltip--top' : placement === 'top',
				'bx--tooltip--align-start' : alignment === 'start',
				'bx--tooltip--align-center' : alignment === 'center',
				'bx--tooltip--align-end' : alignment === 'end'
			}"
			class="bx--tooltip__trigger bx--tooltip--a11y bx--tooltip__trigger--definition"
			[attr.aria-describedby]="id">
			<ng-content></ng-content>
		</button>
		<div class="bx--assistive-text" [id]="id" role="tooltip">{{content}}</div>
	`
            }] }
];
TooltipDefinition.propDecorators = {
    id: [{ type: Input }],
    content: [{ type: Input }],
    placement: [{ type: Input }],
    alignment: [{ type: Input }],
    className: [{ type: HostBinding, args: ["class.bx--tooltip--definition",] }],
    accessibility: [{ type: HostBinding, args: ["class.bx--tooltip--a11y",] }]
};
