// carbon-components-angular/esm5/ui-shell/sidenav/sidenav.module.js
/**
 *
 * carbon-angular v0.0.0 | sidenav.module.js
 *
 * Copyright 2014, 2021 IBM
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0

 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */


import { NgModule } from "@angular/core";
import { CommonModule } from "@angular/common";
import { I18nModule } from "carbon-components-angular/i18n";
import { SideNav } from "./sidenav.component";
import { SideNavItem } from "./sidenav-item.component";
import { SideNavMenu } from "./sidenav-menu.component";
export { SideNav, SideNavItem, SideNavMenu };
var SideNavModule = /** @class */ (function () {
    function SideNavModule() {
    }
    SideNavModule.decorators = [
        { type: NgModule, args: [{
                    declarations: [
                        SideNav,
                        SideNavItem,
                        SideNavMenu
                    ],
                    imports: [CommonModule, I18nModule],
                    exports: [
                        SideNav,
                        SideNavItem,
                        SideNavMenu
                    ]
                },] }
    ];
    return SideNavModule;
}());
export { SideNavModule };
