// carbon-components-angular/esm5/dialog/overflow-menu/overflow-menu.directive.js
/**
 *
 * carbon-angular v0.0.0 | overflow-menu.directive.js
 *
 * Copyright 2014, 2021 IBM
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0

 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */


import * as tslib_1 from "tslib";
import { Directive, ElementRef, ViewContainerRef, Input, TemplateRef, HostListener } from "@angular/core";
import { DialogDirective } from "../dialog.directive";
import { DialogService } from "../dialog.service";
import { OverflowMenuPane } from "./overflow-menu-pane.component";
import { OverflowMenuCustomPane } from "./overflow-menu-custom-pane.component";
import { EventService } from "carbon-components-angular/utils";
/**
 * Directive for extending `Dialog` to create overflow menus.
 *
 * class: OverflowMenuDirective (extends DialogDirective)
 *
 *
 * selector: `ibmOverflowMenu`
 *
 *
 * ```html
 * <div [ibmOverflowMenu]="templateRef"></div>
 * <ng-template #templateRef>
 * 	<!-- overflow menu options here -->
 * </ng-template>
 * ```
 *
 * ```html
 * <div [ibmOverflowMenu]="templateRef" [customPane]="true"></div>
 * <ng-template #templateRef>
 *  <!-- custom content goes here -->
 * </ng-template>
 * ```
 */
var OverflowMenuDirective = /** @class */ (function (_super) {
    tslib_1.__extends(OverflowMenuDirective, _super);
    /**
     * Creates an instance of `OverflowMenuDirective`.
     */
    function OverflowMenuDirective(elementRef, viewContainerRef, dialogService, eventService) {
        var _this = _super.call(this, elementRef, viewContainerRef, dialogService, eventService) || this;
        _this.elementRef = elementRef;
        _this.viewContainerRef = viewContainerRef;
        _this.dialogService = dialogService;
        _this.eventService = eventService;
        /**
         * Controls wether the overflow menu is flipped
         */
        _this.flip = false;
        /**
         * Classes to add to the dialog container
         */
        _this.wrapperClass = "";
        /**
         * Set to true to for custom content
         */
        _this.customPane = false;
        return _this;
    }
    OverflowMenuDirective.prototype.ngAfterContentInit = function () {
        this.dialogService.setContext({ component: this.customPane ? OverflowMenuCustomPane : OverflowMenuPane });
    };
    OverflowMenuDirective.prototype.updateConfig = function () {
        this.dialogConfig.content = this.ibmOverflowMenu;
        this.dialogConfig.flip = this.flip;
        this.dialogConfig.offset = this.offset;
        this.dialogConfig.wrapperClass = this.wrapperClass;
    };
    OverflowMenuDirective.prototype.hostkeys = function (event) {
        switch (event.key) {
            case "Enter":
            case " ":
                event.preventDefault();
                break;
        }
    };
    OverflowMenuDirective.decorators = [
        { type: Directive, args: [{
                    selector: "[ibmOverflowMenu]",
                    exportAs: "ibmOverflowMenu",
                    providers: [
                        DialogService
                    ]
                },] }
    ];
    /** @nocollapse */
    OverflowMenuDirective.ctorParameters = function () { return [
        { type: ElementRef },
        { type: ViewContainerRef },
        { type: DialogService },
        { type: EventService }
    ]; };
    OverflowMenuDirective.propDecorators = {
        ibmOverflowMenu: [{ type: Input }],
        flip: [{ type: Input }],
        offset: [{ type: Input }],
        wrapperClass: [{ type: Input }],
        customPane: [{ type: Input }],
        hostkeys: [{ type: HostListener, args: ["keydown", ["$event"],] }]
    };
    return OverflowMenuDirective;
}(DialogDirective));
export { OverflowMenuDirective };
