// carbon-components-angular/esm2015/table/table.directive.js
/**
 *
 * carbon-angular v0.0.0 | table.directive.js
 *
 * Copyright 2014, 2021 IBM
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0

 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */


import { Directive, HostBinding, Input } from "@angular/core";
export class TableDirective {
    constructor() {
        this.sortable = true;
        this.noBorder = true;
        this.striped = false;
        this.skeleton = false;
        /**
         * Size of the table rows.
         */
        this.size = "md";
        this.tableClass = true;
    }
    get compactClass() {
        return this.size === "sm";
    }
    get tallClass() {
        return this.size === "lg";
    }
    get shortClass() {
        return this.size === "sh";
    }
}
TableDirective.decorators = [
    { type: Directive, args: [{
                selector: "[ibmTable]"
            },] }
];
TableDirective.propDecorators = {
    sortable: [{ type: Input }, { type: HostBinding, args: ["class.bx--data-table--sort",] }],
    noBorder: [{ type: Input }, { type: HostBinding, args: ["class.bx--data-table--no-border",] }],
    striped: [{ type: Input }, { type: HostBinding, args: ["class.bx--data-table--zebra",] }],
    skeleton: [{ type: Input }, { type: HostBinding, args: ["class.bx--skeleton",] }],
    size: [{ type: Input }],
    tableClass: [{ type: HostBinding, args: ["class.bx--data-table",] }],
    compactClass: [{ type: HostBinding, args: ["class.bx--data-table--compact",] }],
    tallClass: [{ type: HostBinding, args: ["class.bx--data-table--tall",] }],
    shortClass: [{ type: HostBinding, args: ["class.bx--data-table--short",] }]
};
