// carbon-components-angular/esm5/table/head/table-head-checkbox.component.js
/**
 *
 * carbon-angular v0.0.0 | table-head-checkbox.component.js
 *
 * Copyright 2014, 2021 IBM
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0

 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */


import { Component, Input, Output, HostBinding, EventEmitter } from "@angular/core";
import { I18n } from "carbon-components-angular/i18n";
var TableHeadCheckbox = /** @class */ (function () {
    function TableHeadCheckbox(i18n) {
        this.i18n = i18n;
        /**
         * Size of the table rows.
         */
        this.size = "md";
        this.checked = false;
        this.indeterminate = false;
        this.skeleton = false;
        this.name = "select-all-" + TableHeadCheckbox.tableSelectAllCount++;
        this.change = new EventEmitter();
        this.hostClass = true;
        this._ariaLabel = this.i18n.getOverridable("TABLE.CHECKBOX_HEADER");
    }
    Object.defineProperty(TableHeadCheckbox.prototype, "ariaLabel", {
        get: function () {
            return this._ariaLabel.value;
        },
        set: function (value) {
            this._ariaLabel.override(value);
        },
        enumerable: true,
        configurable: true
    });
    TableHeadCheckbox.prototype.getAriaLabel = function () {
        return this._ariaLabel.subject;
    };
    TableHeadCheckbox.tableSelectAllCount = 0;
    TableHeadCheckbox.decorators = [
        { type: Component, args: [{
                    // tslint:disable-next-line: component-selector
                    selector: "[ibmTableHeadCheckbox]",
                    template: "\n\t\t<ibm-checkbox\n\t\t\t*ngIf=\"!skeleton\"\n\t\t\tinline=\"true\"\n\t\t\t[size]=\"(size !== 'sm' ? 'md' : 'sm')\"\n\t\t\t[name]=\"name\"\n\t\t\t[checked]=\"checked\"\n\t\t\t[indeterminate]=\"indeterminate\"\n\t\t\t(checkedChange)=\"change.emit()\"\n\t\t\t[aria-label]=\"getAriaLabel() | async\">\n\t\t</ibm-checkbox>\n\t",
                    styles: ["\n        :host { width: 10px; }\n    "]
                }] }
    ];
    /** @nocollapse */
    TableHeadCheckbox.ctorParameters = function () { return [
        { type: I18n }
    ]; };
    TableHeadCheckbox.propDecorators = {
        size: [{ type: Input }],
        checked: [{ type: Input }],
        indeterminate: [{ type: Input }],
        skeleton: [{ type: Input }],
        name: [{ type: Input }],
        ariaLabel: [{ type: Input }],
        change: [{ type: Output }],
        hostClass: [{ type: HostBinding, args: ["class.bx--table-column-checkbox",] }]
    };
    return TableHeadCheckbox;
}());
export { TableHeadCheckbox };
