// carbon-components-angular/esm5/dialog/dialog.module.js
/**
 *
 * carbon-angular v0.0.0 | dialog.module.js
 *
 * Copyright 2014, 2021 IBM
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0

 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */


// modules
import { NgModule } from "@angular/core";
import { CommonModule } from "@angular/common";
// imports
import { DialogService } from "./dialog.service";
import { Dialog } from "./dialog.component";
import { DialogDirective } from "./dialog.directive";
import { Tooltip } from "./tooltip/tooltip.component";
import { TooltipDefinition } from "./tooltip/tooltip-definition.component";
import { TooltipIcon } from "./tooltip/tooltip-icon.component";
import { TooltipDirective } from "./tooltip/tooltip.directive";
import { EllipsisTooltip } from "./tooltip/ellipsis-tooltip.directive";
import { OverflowMenu } from "./overflow-menu/overflow-menu.component";
import { OverflowMenuPane } from "./overflow-menu/overflow-menu-pane.component";
import { OverflowMenuCustomPane } from "./overflow-menu/overflow-menu-custom-pane.component";
import { OverflowMenuDirective } from "./overflow-menu/overflow-menu.directive";
import { OverflowMenuOption } from "./overflow-menu/overflow-menu-option.component";
import { I18nModule } from "carbon-components-angular/i18n";
import { PlaceholderModule } from "carbon-components-angular/placeholder";
import { ExperimentalModule } from "carbon-components-angular/experimental";
import { UtilsModule } from "carbon-components-angular/utils";
import { IconModule } from "carbon-components-angular/icon";
var DialogModule = /** @class */ (function () {
    function DialogModule() {
    }
    DialogModule.decorators = [
        { type: NgModule, args: [{
                    declarations: [
                        Dialog,
                        Tooltip,
                        TooltipDefinition,
                        TooltipIcon,
                        OverflowMenu,
                        OverflowMenuPane,
                        OverflowMenuCustomPane,
                        DialogDirective,
                        TooltipDirective,
                        EllipsisTooltip,
                        OverflowMenuDirective,
                        OverflowMenuOption
                    ],
                    exports: [
                        Dialog,
                        Tooltip,
                        TooltipDefinition,
                        TooltipIcon,
                        OverflowMenu,
                        OverflowMenuPane,
                        OverflowMenuCustomPane,
                        DialogDirective,
                        TooltipDirective,
                        EllipsisTooltip,
                        OverflowMenuDirective,
                        OverflowMenuOption
                    ],
                    providers: [DialogService],
                    entryComponents: [
                        Dialog,
                        Tooltip,
                        OverflowMenuPane,
                        OverflowMenuCustomPane
                    ],
                    imports: [
                        CommonModule,
                        I18nModule,
                        PlaceholderModule,
                        ExperimentalModule,
                        UtilsModule,
                        IconModule
                    ]
                },] }
    ];
    return DialogModule;
}());
export { DialogModule };
