// carbon-components-angular/esm5/table/cell/table-checkbox.component.js
/**
 *
 * carbon-angular v0.0.0 | table-checkbox.component.js
 *
 * Copyright 2014, 2021 IBM
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0

 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */


import { Component, Input, Output, EventEmitter } from "@angular/core";
import { I18n } from "carbon-components-angular/i18n";
var TableCheckbox = /** @class */ (function () {
    function TableCheckbox(i18n) {
        this.i18n = i18n;
        this.selected = false;
        this.name = "";
        /**
         * Size of the table rows.
         */
        this.size = "md";
        this.skeleton = false;
        /**
         * Emits if a single row is selected.
         */
        this.change = new EventEmitter();
        this._label = this.i18n.getOverridable("TABLE.CHECKBOX_ROW");
    }
    Object.defineProperty(TableCheckbox.prototype, "disabled", {
        get: function () {
            return this.row ? !!this.row.disabled : false;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(TableCheckbox.prototype, "label", {
        get: function () {
            return this._label.value;
        },
        set: function (value) {
            this._label.override(value);
        },
        enumerable: true,
        configurable: true
    });
    TableCheckbox.prototype.getSelectionLabelValue = function (row) {
        if (!this.selectionLabelColumn) {
            return { value: this.i18n.get().TABLE.ROW };
        }
        return { value: row[this.selectionLabelColumn].data };
    };
    TableCheckbox.prototype.getLabel = function () {
        return this._label.subject;
    };
    TableCheckbox.decorators = [
        { type: Component, args: [{
                    // tslint:disable-next-line: component-selector
                    selector: "[ibmTableCheckbox]",
                    template: "\n\t\t<ibm-checkbox\n\t\t\t*ngIf=\"!skeleton\"\n\t\t\tinline=\"true\"\n\t\t\t[name]=\"name\"\n\t\t\t[aria-label]=\"getLabel() | i18nReplace:getSelectionLabelValue(row) | async\"\n\t\t\t[size]=\"(size !== 'sm' ? 'md' : 'sm')\"\n\t\t\t[checked]=\"selected\"\n\t\t\t[disabled]=\"disabled\"\n\t\t\t(change)=\"change.emit()\">\n\t\t</ibm-checkbox>\n\t"
                }] }
    ];
    /** @nocollapse */
    TableCheckbox.ctorParameters = function () { return [
        { type: I18n }
    ]; };
    TableCheckbox.propDecorators = {
        row: [{ type: Input }],
        selected: [{ type: Input }],
        name: [{ type: Input }],
        size: [{ type: Input }],
        label: [{ type: Input }],
        selectionLabelColumn: [{ type: Input }],
        skeleton: [{ type: Input }],
        change: [{ type: Output }]
    };
    return TableCheckbox;
}());
export { TableCheckbox };
