// carbon-components-angular/esm5/input/input.directive.js
/**
 *
 * carbon-angular v0.0.0 | input.directive.js
 *
 * Copyright 2014, 2021 IBM
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0

 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */


import { Directive, HostBinding, Input } from "@angular/core";
/**
 * A directive for applying styling to an input element.
 *
 * Example:
 *
 * ```html
 * <input ibmText/>
 * ```
 *
 * See the [vanilla carbon docs](http://www.carbondesignsystem.com/components/text-input/code) for more detail.
 */
var TextInput = /** @class */ (function () {
    function TextInput() {
        /**
         * `light` or `dark` input theme
         */
        this.theme = "dark";
        /**
         * Input field render size
         */
        this.size = "md";
        this.inputClass = true;
        this.invalid = false;
        this.skeleton = false;
    }
    Object.defineProperty(TextInput.prototype, "isSizeXl", {
        get: function () {
            return this.size === "xl";
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(TextInput.prototype, "isSizeSm", {
        get: function () {
            return this.size === "sm";
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(TextInput.prototype, "isLightTheme", {
        get: function () {
            return this.theme === "light";
        },
        enumerable: true,
        configurable: true
    });
    TextInput.decorators = [
        { type: Directive, args: [{
                    selector: "[ibmText]"
                },] }
    ];
    TextInput.propDecorators = {
        theme: [{ type: Input }],
        size: [{ type: Input }],
        inputClass: [{ type: HostBinding, args: ["class.bx--text-input",] }],
        isSizeXl: [{ type: HostBinding, args: ["class.bx--text-input--xl",] }],
        isSizeSm: [{ type: HostBinding, args: ["class.bx--text-input--sm",] }],
        invalid: [{ type: HostBinding, args: ["class.bx--text-input--invalid",] }, { type: Input }],
        skeleton: [{ type: HostBinding, args: ["class.bx--skeleton",] }, { type: Input }],
        isLightTheme: [{ type: HostBinding, args: ["class.bx--text-input--light",] }]
    };
    return TextInput;
}());
export { TextInput };
