// carbon-components-angular/esm2015/modal/index.js
/**
 *
 * carbon-angular v0.0.0 | index.js
 *
 * Copyright 2014, 2021 IBM
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0

 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */


export { AlertModal } from "./alert-modal.component";
export { AlertModalType, ModalButtonType } from "./alert-modal.interface";
export { BaseModal } from "./base-modal.class";
export { ModalFooter } from "./modal-footer.component";
export { ModalHeader } from "./modal-header.component";
export { Modal } from "./modal.component";
export { ModalModule } from "./modal.module";
export { ModalService } from "./modal.service";
export { Overlay } from "./overlay.component";
export { ModalContent } from "./modal-content.directive";
export { ModalContentText } from "./modal-content-text.directive";
export { ModalHeaderHeading } from "./modal-header-heading.directive";
export { ModalHeaderLabel } from "./modal-header-label.directive";
