// carbon-components-angular/esm5/table/table.directive.js
/**
 *
 * carbon-angular v0.0.0 | table.directive.js
 *
 * Copyright 2014, 2021 IBM
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0

 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */


import { Directive, HostBinding, Input } from "@angular/core";
var TableDirective = /** @class */ (function () {
    function TableDirective() {
        this.sortable = true;
        this.noBorder = true;
        this.striped = false;
        this.skeleton = false;
        /**
         * Size of the table rows.
         */
        this.size = "md";
        this.tableClass = true;
    }
    Object.defineProperty(TableDirective.prototype, "compactClass", {
        get: function () {
            return this.size === "sm";
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(TableDirective.prototype, "tallClass", {
        get: function () {
            return this.size === "lg";
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(TableDirective.prototype, "shortClass", {
        get: function () {
            return this.size === "sh";
        },
        enumerable: true,
        configurable: true
    });
    TableDirective.decorators = [
        { type: Directive, args: [{
                    selector: "[ibmTable]"
                },] }
    ];
    TableDirective.propDecorators = {
        sortable: [{ type: Input }, { type: HostBinding, args: ["class.bx--data-table--sort",] }],
        noBorder: [{ type: Input }, { type: HostBinding, args: ["class.bx--data-table--no-border",] }],
        striped: [{ type: Input }, { type: HostBinding, args: ["class.bx--data-table--zebra",] }],
        skeleton: [{ type: Input }, { type: HostBinding, args: ["class.bx--skeleton",] }],
        size: [{ type: Input }],
        tableClass: [{ type: HostBinding, args: ["class.bx--data-table",] }],
        compactClass: [{ type: HostBinding, args: ["class.bx--data-table--compact",] }],
        tallClass: [{ type: HostBinding, args: ["class.bx--data-table--tall",] }],
        shortClass: [{ type: HostBinding, args: ["class.bx--data-table--short",] }]
    };
    return TableDirective;
}());
export { TableDirective };
