// carbon-components-angular/esm5/table/table-adapter.class.js
/**
 *
 * carbon-angular v0.0.0 | table-adapter.class.js
 *
 * Copyright 2014, 2021 IBM
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0

 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */


import * as tslib_1 from "tslib";
/**
 * An abstract class that represents a cell in a table
 */
var TableCellAdapter = /** @class */ (function () {
    function TableCellAdapter() {
    }
    return TableCellAdapter;
}());
export { TableCellAdapter };
/**
 * An abstract class that represents a row in a table
 */
var TableRowAdapter = /** @class */ (function () {
    function TableRowAdapter() {
    }
    return TableRowAdapter;
}());
export { TableRowAdapter };
/**
 * An abstract representation of a table that provides
 * a standard interface to query 2d tables for cell and row information.
 */
var TableAdapter = /** @class */ (function () {
    function TableAdapter() {
    }
    Object.defineProperty(TableAdapter.prototype, "lastColumnIndex", {
        /**
         * The last accessible column in the table
         */
        get: function () { return; },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(TableAdapter.prototype, "lastRowIndex", {
        /**
         * The last accessible row in the table
         */
        get: function () { return; },
        enumerable: true,
        configurable: true
    });
    /**
     * Returns a cell from the table
     *
     * @param row index of the row
     * @param column index of the column
     */
    TableAdapter.prototype.getCell = function (row, column) { return; };
    /**
     * Returns a row from the table
     *
     * @param column index of the column
     */
    TableAdapter.prototype.getColumn = function (column) { return; };
    /**
     * Returns a row from the table
     *
     * @param row index of the row
     */
    TableAdapter.prototype.getRow = function (row) { return; };
    /**
     * Finds the column index of a given cell
     *
     * @param cell the cell to search for
     */
    TableAdapter.prototype.findColumnIndex = function (cell) { return; };
    /**
     * Finds the row index of a given cell
     *
     * @param cell the cell to search for
     */
    TableAdapter.prototype.findRowIndex = function (cell) { return; };
    /**
     * Finds the row and column index of a given cell
     *
     * @param cell the cell to search for
     * @returns a tuple that follows the `[row, column]` convention
     */
    TableAdapter.prototype.findIndex = function (cell) { return; };
    return TableAdapter;
}());
export { TableAdapter };
var TableDomSpanDirection;
(function (TableDomSpanDirection) {
    TableDomSpanDirection["colSpan"] = "colSpan";
    TableDomSpanDirection["rowSpan"] = "rowSpan";
})(TableDomSpanDirection || (TableDomSpanDirection = {}));
/**
 * A concrete implementation of `TableAdapter`
 *
 * Provides standard and consistent access to table cells and rows
 */
var TableDomAdapter = /** @class */ (function () {
    /**
     * `TableDomAdapter` works on a normal HTML table structure.
     * Custom tables that don't follow the standard structure should use a custom implementation of `TableAdapter`.
     *
     * The standard structure allows us to directly query rows for cells and indexes - though we do have to handle colspans specially.
     *
     * @param tableElement the root HTML table element.
     */
    function TableDomAdapter(tableElement) {
        this.tableElement = tableElement;
    }
    Object.defineProperty(TableDomAdapter.prototype, "lastColumnIndex", {
        /**
         * The last accessible column in the table
         */
        get: function () {
            return this.getRealRowLength(this.tableElement.rows[0]);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(TableDomAdapter.prototype, "lastRowIndex", {
        /**
         * The last accessible row in the table
         */
        get: function () {
            return this.tableElement.rows.length - 1;
        },
        enumerable: true,
        configurable: true
    });
    /**
     * Returns a cell from the table taking colspans in to account.
     *
     * @param row index of the row
     * @param column index of the column
     */
    TableDomAdapter.prototype.getCell = function (row, column) {
        var col = this.getColumn(column);
        return this.findCellInColumn(col, row).cell;
    };
    /**
     * Returns a column from the table, using the `id` and `headers` attributes
     *
     * See here for more detail these attributes: https://www.w3.org/TR/WCAG20-TECHS/H43.html
     *
     * @param column the index of the column
     */
    TableDomAdapter.prototype.getColumn = function (column) {
        var firstHeader = Array.from(this.tableElement.rows[0].cells);
        var _a = this.findCellInRow(firstHeader, column), header = _a.cell, realColumnIndex = _a.realIndex;
        var linkedCells = [];
        for (var i = 1; i < this.tableElement.rows.length; i++) {
            var row = this.tableElement.rows[i];
            // query for any cells that are linked to the given header id
            // `~=` matches values in space separated lists - so `[headers~='foo']` would match `headers="foo bar"` and `headers="foo"`
            // but not `headers="bar"` or `headers="bar baz"`
            var linkedRowCells = row.querySelectorAll("[headers~='" + header.id + "']");
            // if we have more than one cell, get the one that is closest to the column
            if (linkedRowCells.length > 1) {
                var cell = this.findCellInRow(Array.from(linkedRowCells), column - realColumnIndex).cell;
                linkedCells.push(cell);
            }
            else if (linkedRowCells[0]) {
                linkedCells.push(linkedRowCells[0]);
            }
        }
        // return an empty array if we can't find any linked cells
        // returning anything else would be a lie
        if (!linkedCells) {
            return [];
        }
        return tslib_1.__spread([header], linkedCells);
    };
    /**
     * Returns a row from the table
     *
     * @param row index of the row
     */
    TableDomAdapter.prototype.getRow = function (row) {
        return this.tableElement.rows[row];
    };
    /**
     * Finds the column index of a given cell
     *
     * @param cell the cell to search for
     */
    TableDomAdapter.prototype.findColumnIndex = function (cell) {
        var e_1, _a, e_2, _b, e_3, _c, e_4, _d, e_5, _e;
        var row = this.getRow(this.findRowIndex(cell));
        if (!row) {
            return;
        }
        // if the cell has linked headers we can do a more accurate lookup
        if (cell && cell.headers) {
            var ids_2 = cell.headers.split(" ");
            var headerRows = Array.from(this.tableElement.tHead.rows);
            var indexes = [];
            try {
                // start from the last row and work up
                for (var _f = tslib_1.__values(headerRows.reverse()), _g = _f.next(); !_g.done; _g = _f.next()) {
                    var headerRow = _g.value;
                    var headerCells = Array.from(headerRow.cells);
                    var header = headerCells.find(function (headerCell) { return ids_2.includes(headerCell.id); });
                    // if we have a matching header, find it's index (adjusting for colspans)
                    if (header) {
                        // this is borrowed from below
                        var cellIndex_1 = 0;
                        try {
                            for (var headerCells_1 = tslib_1.__values(headerCells), headerCells_1_1 = headerCells_1.next(); !headerCells_1_1.done; headerCells_1_1 = headerCells_1.next()) {
                                var c = headerCells_1_1.value;
                                if (c === header) {
                                    break;
                                }
                                cellIndex_1 += c.colSpan;
                            }
                        }
                        catch (e_2_1) { e_2 = { error: e_2_1 }; }
                        finally {
                            try {
                                if (headerCells_1_1 && !headerCells_1_1.done && (_b = headerCells_1.return)) _b.call(headerCells_1);
                            }
                            finally { if (e_2) throw e_2.error; }
                        }
                        indexes.push(cellIndex_1);
                    }
                }
            }
            catch (e_1_1) { e_1 = { error: e_1_1 }; }
            finally {
                try {
                    if (_g && !_g.done && (_a = _f.return)) _a.call(_f);
                }
                finally { if (e_1) throw e_1.error; }
            }
            // sort the indexes largest to smallest to find the closest matching header index
            var firstIndex = indexes.sort(function (a, b) { return b - a; })[0];
            // search the row for cells that share the header
            var similarCells = [];
            try {
                for (var ids_1 = tslib_1.__values(ids_2), ids_1_1 = ids_1.next(); !ids_1_1.done; ids_1_1 = ids_1.next()) {
                    var id = ids_1_1.value;
                    // there's no selector that will match two space separated lists,
                    // so we have to iterate through the ids and query the row for each
                    var rowCells = Array.from(row.querySelectorAll("[headers~='" + id + "']"));
                    try {
                        for (var rowCells_1 = tslib_1.__values(rowCells), rowCells_1_1 = rowCells_1.next(); !rowCells_1_1.done; rowCells_1_1 = rowCells_1.next()) {
                            var rowCell = rowCells_1_1.value;
                            // only keep one set of cells
                            if (!similarCells.includes(rowCell)) {
                                similarCells.push(rowCell);
                            }
                        }
                    }
                    catch (e_4_1) { e_4 = { error: e_4_1 }; }
                    finally {
                        try {
                            if (rowCells_1_1 && !rowCells_1_1.done && (_d = rowCells_1.return)) _d.call(rowCells_1);
                        }
                        finally { if (e_4) throw e_4.error; }
                    }
                }
            }
            catch (e_3_1) { e_3 = { error: e_3_1 }; }
            finally {
                try {
                    if (ids_1_1 && !ids_1_1.done && (_c = ids_1.return)) _c.call(ids_1);
                }
                finally { if (e_3) throw e_3.error; }
            }
            // DOM order is not preserved, so we have to sort the row
            similarCells = similarCells.sort(function (a, b) { return a.cellIndex - b.cellIndex; });
            // return the header index plus any adjustment within that headers column
            return firstIndex + similarCells.indexOf(cell);
        }
        // fallback if the cell isn't linked to any headers
        var cellIndex = 0;
        try {
            for (var _h = tslib_1.__values(Array.from(row.cells)), _j = _h.next(); !_j.done; _j = _h.next()) {
                var c = _j.value;
                if (c === cell) {
                    break;
                }
                cellIndex += c.colSpan;
            }
        }
        catch (e_5_1) { e_5 = { error: e_5_1 }; }
        finally {
            try {
                if (_j && !_j.done && (_e = _h.return)) _e.call(_h);
            }
            finally { if (e_5) throw e_5.error; }
        }
        return cellIndex;
    };
    /**
     * Finds the row index of a given cell
     *
     * @param cell the cell to search for
     */
    TableDomAdapter.prototype.findRowIndex = function (cell) {
        var e_6, _a;
        try {
            for (var _b = tslib_1.__values(Array.from(this.tableElement.rows)), _c = _b.next(); !_c.done; _c = _b.next()) {
                var row = _c.value;
                if (row.contains(cell)) {
                    return row.rowIndex;
                }
            }
        }
        catch (e_6_1) { e_6 = { error: e_6_1 }; }
        finally {
            try {
                if (_c && !_c.done && (_a = _b.return)) _a.call(_b);
            }
            finally { if (e_6) throw e_6.error; }
        }
    };
    /**
     * Finds the row and column index of a given cell
     *
     * @param cell the cell to search for
     * @returns a tuple that follows the `[row, column]` convention
     */
    TableDomAdapter.prototype.findIndex = function (cell) {
        return [this.findRowIndex(cell), this.findColumnIndex(cell)];
    };
    /**
     * Helper function that returns the "real" length of a row.
     * Only accurate with regard to colspans (though that's sufficient for it's uses here)
     *
     * TODO: Take rowSpan into account
     *
     * @param row the row to get the length of
     */
    TableDomAdapter.prototype.getRealRowLength = function (row) {
        // start at -1 since the colspans will sum to 1 index greater than the total
        return Array.from(row.cells).reduce(function (count, cell) { return count + cell.colSpan; }, -1);
    };
    /**
     * Finds a cell and it's real index given an array of cells, a target index, and the spanning direction
     *
     * @param cells An array of cells to search
     * @param targetIndex The index we think the cell is located at
     * @param spanDirection The direction of the cell spans. Should be `"colSpan"` for a row and `"rowSpan"` for a column
     */
    TableDomAdapter.prototype.findCell = function (cells, targetIndex, spanDirection) {
        // rows/cols can have fewer total cells than the actual table
        // the model pretends all rows/cols behave the same (with col/row spans > 1 being N cells long)
        // this maps that view to the HTML view (col/row spans > 1 are one element, so the array is shorter)
        var realIndex = 0;
        // i is only used for iterating the cells
        for (var i = 0; i < targetIndex;) {
            // skip the next N cells
            i += cells[realIndex][spanDirection];
            // don't bump realIndex if i now exceeds the cell we're shooting for
            if (i > targetIndex) {
                break;
            }
            // finally, increment realIndex (to keep it generally in step with i)
            realIndex++;
        }
        return {
            cell: cells[realIndex],
            realIndex: realIndex
        };
    };
    /**
     * Helper method around `findCell`, searches based on a row of cells
     *
     * @param row the row of elements to search
     * @param index the index of the element
     */
    TableDomAdapter.prototype.findCellInRow = function (row, index) {
        return this.findCell(row, index, TableDomSpanDirection.colSpan);
    };
    /**
     * Helper method around `findCell`, searches based on a column of cells
     *
     * @param col the column of elements to search
     * @param index the index of the element
     */
    TableDomAdapter.prototype.findCellInColumn = function (col, index) {
        return this.findCell(col, index, TableDomSpanDirection.rowSpan);
    };
    return TableDomAdapter;
}());
export { TableDomAdapter };
