// carbon-components-angular/esm5/pagination/pagination-nav/pagination-item.component.js
/**
 *
 * carbon-angular v0.0.0 | pagination-item.component.js
 *
 * Copyright 2014, 2021 IBM
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0

 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */


import { Component, Input, Output, EventEmitter } from "@angular/core";
/**
 * Used to present a single navigation item in a pagination list
 *
 *  * ```html
 * <ibm-pagination-nav-item [page]="5" [isActive]="false" (click)="handleClick(value)"></ibm-pagination-nav-item>
 * ```
 */
var PaginationNavItem = /** @class */ (function () {
    function PaginationNavItem() {
        /**
         * The page for this component to dipslay
         */
        this.page = 0;
        /**
         * The state for this component to dipslay
         */
        this.isActive = false;
        /**
         * Emits click event
         */
        this.click = new EventEmitter();
    }
    PaginationNavItem.decorators = [
        { type: Component, args: [{
                    selector: "ibm-pagination-nav-item",
                    template: "\n\t\t<li class=\"bx--pagination-nav__list-item\">\n\t\t\t<button\n\t\t\t\ttype=\"button\"\n\t\t\t\tclass=\"bx--pagination-nav__page\"\n\t\t\t\t[ngClass]=\"{ 'bx--pagination-nav__page--active': isActive }\"\n\t\t\t\t(click)=\"click.emit(page)\">\n\t\t\t\t<span class=\"bx--pagination-nav__accessibility-label\">\n\t\t\t\t\t{{page}}\n\t\t\t\t</span>\n\t\t\t\t{{page}}\n\t\t\t</button>\n\t\t</li>\n\t"
                }] }
    ];
    /** @nocollapse */
    PaginationNavItem.ctorParameters = function () { return []; };
    PaginationNavItem.propDecorators = {
        page: [{ type: Input }],
        isActive: [{ type: Input }],
        click: [{ type: Output }]
    };
    return PaginationNavItem;
}());
export { PaginationNavItem };
