// carbon-components-angular/bundles/carbon-components-angular-link.umd.js
/**
 *
 * carbon-angular v0.0.0 | carbon-components-angular-link.umd.js
 *
 * Copyright 2014, 2021 IBM
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0

 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */


(function (global, factory) {
    typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports, require('@angular/core'), require('@angular/common')) :
    typeof define === 'function' && define.amd ? define('carbon-components-angular/link', ['exports', '@angular/core', '@angular/common'], factory) :
    (factory((global['carbon-components-angular'] = global['carbon-components-angular'] || {}, global['carbon-components-angular'].link = {}),global.ng.core,global.ng.common));
}(this, (function (exports,core,common) { 'use strict';

    /**
     * A convenience directive for applying styling to a link.
     *
     * [See demo](../../?path=/story/link--basic)
     *
     * Example:
     *
     * ```html
     * <a href="#" ibmLink>A link</a>
     * ```
     *
     * See the [vanilla carbon docs](http://www.carbondesignsystem.com/components/link/code) for more detail.
     *
     * <example-url>../../iframe.html?id=link--basic</example-url>
     */
    var Link = /** @class */ (function () {
        function Link() {
            this.baseClass = true;
            /**
             * Set to true to show links inline in a sentence or paragraph.
             */
            this.inline = false;
        }
        Object.defineProperty(Link.prototype, "disabled", {
            get: function () {
                return this._disabled;
            },
            /**
             * Set to true to disable link.
             */
            set: function (disabled) {
                this._disabled = disabled;
                this.tabindex = this.disabled ? -1 : null;
            },
            enumerable: true,
            configurable: true
        });
        Link.decorators = [
            { type: core.Directive, args: [{
                        selector: "[ibmLink]"
                    },] }
        ];
        Link.propDecorators = {
            baseClass: [{ type: core.HostBinding, args: ["class.bx--link",] }],
            tabindex: [{ type: core.HostBinding, args: ["attr.tabindex",] }],
            inline: [{ type: core.Input }, { type: core.HostBinding, args: ["class.bx--link--inline",] }],
            disabled: [{ type: core.Input }, { type: core.HostBinding, args: ["attr.aria-disabled",] }, { type: core.HostBinding, args: ["class.bx--link--disabled",] }]
        };
        return Link;
    }());

    var LinkModule = /** @class */ (function () {
        function LinkModule() {
        }
        LinkModule.decorators = [
            { type: core.NgModule, args: [{
                        declarations: [
                            Link
                        ],
                        exports: [
                            Link
                        ],
                        imports: [
                            common.CommonModule
                        ]
                    },] }
        ];
        return LinkModule;
    }());

    /**
     * Generated bundle index. Do not edit.
     */

    exports.Link = Link;
    exports.LinkModule = LinkModule;

    Object.defineProperty(exports, '__esModule', { value: true });

})));
