// carbon-components-angular/bundles/carbon-components-angular-radio.umd.js
/**
 *
 * carbon-angular v0.0.0 | carbon-components-angular-radio.umd.js
 *
 * Copyright 2014, 2021 IBM
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0

 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */


(function (global, factory) {
    typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports, require('@angular/core'), require('@angular/forms'), require('@angular/common')) :
    typeof define === 'function' && define.amd ? define('carbon-components-angular/radio', ['exports', '@angular/core', '@angular/forms', '@angular/common'], factory) :
    (factory((global['carbon-components-angular'] = global['carbon-components-angular'] || {}, global['carbon-components-angular'].radio = {}),global.ng.core,global.ng.forms,global.ng.common));
}(this, (function (exports,core,forms,common) { 'use strict';

    /**
     * Used to emit changes performed on a `Radio`.
     */
    var RadioChange = /** @class */ (function () {
        function RadioChange(source, value) {
            this.source = source;
            this.value = value;
        }
        return RadioChange;
    }());

    /**
     * class: Radio (extends Checkbox)
     *
     * selector: `n-radio`
     *
     * source: `src/forms/radio.component.ts`
     *
     * ```html
     * <ibm-radio [(ngModel)]="radioState">Radio</ibm-radio>
     * ```
     *
     * Also see: [`RadioGroup`](#ibm-radio-group)
     */
    var Radio = /** @class */ (function () {
        function Radio() {
            this.checked = false;
            this.name = "";
            this.disabled = false;
            this.labelPlacement = "right";
            /**
             * Used to set the `aria-label` attribute on the input label.
             */
            this.ariaLabel = "";
            /**
             * Sets the HTML required attribute
             */
            this.required = false;
            /**
             * The value of the `Radio`.
             */
            this.value = "";
            /**
             * Set to `true` for a loading table.
             */
            this.skeleton = false;
            /**
             * The id for the `Radio`.
             */
            this.id = "radio-" + Radio.radioCount++;
            /**
             * emits when the state of the radio changes
             */
            this.change = new core.EventEmitter();
            this.hostClass = true;
            this._labelledby = "";
            /**
             * Handler provided by the `RadioGroup` to bubble events up
             */
            this.radioChangeHandler = function (event) { };
        }
        Object.defineProperty(Radio.prototype, "ariaLabelledby", {
            get: function () {
                if (this._labelledby) {
                    return this._labelledby;
                }
                return "label-" + this.id;
            },
            set: function (value) {
                this._labelledby = value;
            },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(Radio.prototype, "labelLeft", {
            get: function () {
                return this.labelPlacement === "left";
            },
            enumerable: true,
            configurable: true
        });
        /**
         * Synchronizes with the `RadioGroup` in the event of a changed `Radio`.
         * Emits the changes of both the `RadioGroup` and `Radio`.
         */
        Radio.prototype.onChange = function (event) {
            event.stopPropagation();
        };
        Radio.prototype.onClick = function (event) {
            this.checked = event.target.checked;
            var radioEvent = new RadioChange(this, this.value);
            this.change.emit(radioEvent);
            this.radioChangeHandler(radioEvent);
        };
        /**
         * Method called by `RadioGroup` with a callback function to bubble `RadioChange` events
         * @param fn callback that expects a `RadioChange` as an argument
         */
        Radio.prototype.registerRadioChangeHandler = function (fn) {
            this.radioChangeHandler = fn;
        };
        /**
         * Used to dynamically create unique ids for the `Radio`.
         */
        Radio.radioCount = 0;
        Radio.decorators = [
            { type: core.Component, args: [{
                        selector: "ibm-radio",
                        template: "\n\t\t<input\n\t\t\t*ngIf=\"!skeleton\"\n\t\t\tclass=\"bx--radio-button\"\n\t\t\ttype=\"radio\"\n\t\t\t[checked]=\"checked\"\n\t\t\t[disabled]=\"disabled\"\n\t\t\t[name]=\"name\"\n\t\t\t[id]=\"id\"\n\t\t\t[required]=\"required\"\n\t\t\t[value]=\"value\"\n\t\t\t[attr.aria-labelledby]=\"ariaLabelledby\"\n\t\t\t(change)=\"onChange($event)\"\n\t\t\t(click)=\"onClick($event)\">\n\t\t<div *ngIf=\"skeleton\" class=\"bx--radio-button bx--skeleton\"></div>\n\t\t<label\n\t\t\tclass=\"bx--radio-button__label\"\n\t\t\t[attr.aria-label]=\"ariaLabel\"\n\t\t\t[ngClass]=\"{\n\t\t\t\t'bx--skeleton': skeleton\n\t\t\t}\"\n\t\t\t[for]=\"id\"\n\t\t\tid=\"label-{{id}}\">\n\t\t\t<span class=\"bx--radio-button__appearance\"></span>\n\t\t\t<ng-content></ng-content>\n\t\t</label>\n\t",
                        providers: [
                            {
                                provide: forms.NG_VALUE_ACCESSOR,
                                useExisting: Radio,
                                multi: true
                            }
                        ]
                    }] }
        ];
        Radio.propDecorators = {
            checked: [{ type: core.Input }],
            name: [{ type: core.Input }],
            disabled: [{ type: core.Input }],
            labelPlacement: [{ type: core.Input }],
            ariaLabelledby: [{ type: core.Input }],
            ariaLabel: [{ type: core.Input }],
            required: [{ type: core.Input }],
            value: [{ type: core.Input }],
            skeleton: [{ type: core.Input }],
            id: [{ type: core.Input }],
            change: [{ type: core.Output }],
            hostClass: [{ type: core.HostBinding, args: ["class.bx--radio-button-wrapper",] }],
            labelLeft: [{ type: core.HostBinding, args: ["class.bx--radio-button-wrapper--label-left",] }]
        };
        return Radio;
    }());

    /**
     * [See demo](../../?path=/story/radio--basic)
     *
     * class: RadioGroup
     *
     * selector: `ibm-radio-group`
     *
     * source: `src/forms/radio.component.ts`
     *
     *
     * Ex:
     * ```html
     * <ibm-radio-group [(ngModel)]="radio">
     * 	<ibm-radio *ngFor="let one of manyRadios" [value]="one">
     *		Radio {{one}}
     * 	</ibm-radio>
     * </ibm-radio-group>
     *
     * Radio selected: {{radio}}
     * ```
     *
     * ```typescript
     * manyRadios = ["one", "two", "three", "four", "five", "six"];
     * ```
     *
     * Also see: [`Radio`](#ibm-radio)
     *
     * <example-url>../../iframe.html?id=radio--basic</example-url>
     */
    var RadioGroup = /** @class */ (function () {
        function RadioGroup() {
            this.orientation = "horizontal";
            this.labelPlacement = "right";
            /**
             * Emits event notifying other classes of a change using a `RadioChange` class.
             */
            this.change = new core.EventEmitter();
            /**
             * Set to true to disable the whole radio group
             */
            this.disabled = false;
            /**
             * Binds 'bx--form-item' value to the class for `RadioGroup`.
             */
            this.radioButtonGroupClass = true;
            /**
             * To track whether the `RadioGroup` has been initialized.
             */
            this.isInitialized = false;
            /**
             * Reflects whether or not the input is disabled and cannot be selected.
             */
            this._disabled = false;
            /**
             * Reflects whether or not the dropdown is loading.
             */
            this._skeleton = false;
            /**
             * The value of the selected option within the `RadioGroup`.
             */
            this._value = null;
            /**
             * The `Radio` within the `RadioGroup` that is selected.
             */
            this._selected = null;
            /**
             * The name attribute associated with the `RadioGroup`.
             */
            this._name = "radio-group-" + RadioGroup.radioGroupCount++;
            /**
             * Needed to properly implement ControlValueAccessor.
             */
            this.onTouched = function () { };
            /**
             * Method set in registerOnChange to propagate changes back to the form.
             */
            this.propagateChange = function (_) { };
        }
        Object.defineProperty(RadioGroup.prototype, "selected", {
            /**
             * Returns the `Radio` that is selected within the `RadioGroup`.
             */
            get: function () {
                return this._selected;
            },
            /**
             * Sets the passed in `Radio` item as the selected input within the `RadioGroup`.
             */
            set: function (selected) {
                this._selected = selected;
                this.value = selected ? selected.value : null;
                this.checkSelectedRadio();
            },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(RadioGroup.prototype, "value", {
            /**
             * Returns the value/state of the selected `Radio` within the `RadioGroup`.
             */
            get: function () {
                return this._value;
            },
            /**
             * Sets the value/state of the selected `Radio` within the `RadioGroup` to the passed in value.
             */
            set: function (newValue) {
                if (this._value !== newValue) {
                    this._value = newValue;
                    this.updateSelectedRadioFromValue();
                    this.checkSelectedRadio();
                }
            },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(RadioGroup.prototype, "name", {
            /**
             * Returns the associated name of the `RadioGroup`.
             */
            get: function () {
                return this._name;
            },
            /**
             * Replaces the name associated with the `RadioGroup` with the provided parameter.
             */
            set: function (name) {
                this._name = name;
                this.updateRadios();
            },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(RadioGroup.prototype, "skeleton", {
            /**
             * Returns the skeleton value in the `RadioGroup` if there is one.
             */
            get: function () {
                return this._skeleton;
            },
            /**
             * Sets the skeleton value for all `Radio` to the skeleton value of `RadioGroup`.
             */
            set: function (value) {
                this._skeleton = value;
                this.updateChildren();
            },
            enumerable: true,
            configurable: true
        });
        /**
         * Updates the selected `Radio` to be checked (selected).
         */
        RadioGroup.prototype.checkSelectedRadio = function () {
            if (this.selected && !this._selected.checked) {
                this.selected.checked = true;
            }
        };
        /**
         * Use the value of the `RadioGroup` to update the selected radio to the right state (selected state).
         */
        RadioGroup.prototype.updateSelectedRadioFromValue = function () {
            var _this = this;
            var alreadySelected = this._selected != null && this._selected.value === this._value;
            if (this.radios && !alreadySelected) {
                this._selected = null;
                this.radios.forEach(function (radio) {
                    if (radio.checked || radio.value === _this._value) {
                        _this._selected = radio;
                    }
                });
            }
        };
        /**
         * Creates a class of `RadioChange` to emit the change in the `RadioGroup`.
         */
        RadioGroup.prototype.emitChangeEvent = function (event) {
            this.change.emit(event);
            this.propagateChange(event.value);
            this.onTouched();
        };
        /**
         * Synchronizes radio properties.
         */
        RadioGroup.prototype.updateRadios = function () {
            var _this = this;
            if (this.radios) {
                setTimeout(function () {
                    _this.radios.forEach(function (radio) { return radio.name = _this.name; });
                    if (_this.labelPlacement === "left") {
                        _this.radios.forEach(function (radio) { return radio.labelPlacement = "left"; });
                    }
                });
            }
        };
        /**
         * Updates the value of the `RadioGroup` using the provided parameter.
         */
        RadioGroup.prototype.writeValue = function (value) {
            var _this = this;
            this.value = value;
            setTimeout(function () {
                _this.updateSelectedRadioFromValue();
                _this.checkSelectedRadio();
            });
        };
        RadioGroup.prototype.ngAfterContentInit = function () {
            var _this = this;
            this.radios.changes.subscribe(function () {
                _this.updateRadios();
                _this.updateRadioChangeHandler();
            });
            this.updateChildren();
            this.updateRadioChangeHandler();
        };
        RadioGroup.prototype.ngAfterViewInit = function () {
            this.updateRadios();
        };
        /**
         * Used to set method to propagate changes back to the form.
         */
        RadioGroup.prototype.registerOnChange = function (fn) {
            this.propagateChange = fn;
        };
        /**
         * Registers a callback to be triggered when the control has been touched.
         * @param fn Callback to be triggered when the checkbox is touched.
         */
        RadioGroup.prototype.registerOnTouched = function (fn) {
            this.onTouched = fn;
        };
        RadioGroup.prototype.updateChildren = function () {
            var _this = this;
            if (this.radios) {
                this.radios.forEach(function (child) { return child.skeleton = _this.skeleton; });
            }
        };
        RadioGroup.prototype.updateRadioChangeHandler = function () {
            var _this = this;
            this.radios.forEach(function (radio) {
                radio.registerRadioChangeHandler(function (event) {
                    // update selected and value from the event
                    _this._selected = event.source;
                    _this._value = event.value;
                    // bubble the event
                    _this.emitChangeEvent(event);
                });
            });
        };
        /**
         * Used for creating the `RadioGroup` 'name' property dynamically.
         */
        RadioGroup.radioGroupCount = 0;
        RadioGroup.decorators = [
            { type: core.Component, args: [{
                        selector: "ibm-radio-group",
                        template: "\n\t\t<div\n\t\t\tclass=\"bx--radio-button-group\"\n\t\t\t[attr.aria-label]=\"ariaLabel\"\n\t\t\t[attr.aria-labelledby]=\"ariaLabelledby\"\n\t\t\t[ngClass]=\"{\n\t\t\t\t'bx--radio-button-group--vertical': orientation === 'vertical',\n\t\t\t\t'bx--radio-button-group--label-left': orientation === 'vertical' && labelPlacement === 'left'\n\t\t\t}\">\n\t\t\t<ng-content></ng-content>\n\t\t</div>\n\t",
                        providers: [
                            {
                                provide: forms.NG_VALUE_ACCESSOR,
                                useExisting: RadioGroup,
                                multi: true
                            }
                        ]
                    }] }
        ];
        RadioGroup.propDecorators = {
            orientation: [{ type: core.Input }],
            labelPlacement: [{ type: core.Input }],
            ariaLabel: [{ type: core.Input }],
            ariaLabelledby: [{ type: core.Input }],
            change: [{ type: core.Output }],
            radios: [{ type: core.ContentChildren, args: [core.forwardRef(function () { return Radio; }),] }],
            selected: [{ type: core.Input }],
            value: [{ type: core.Input }],
            name: [{ type: core.Input }],
            disabled: [{ type: core.Input }],
            skeleton: [{ type: core.Input }],
            radioButtonGroupClass: [{ type: core.HostBinding, args: ["class.bx--form-item",] }]
        };
        return RadioGroup;
    }());

    // modules
    var RadioModule = /** @class */ (function () {
        function RadioModule() {
        }
        RadioModule.decorators = [
            { type: core.NgModule, args: [{
                        declarations: [
                            Radio,
                            RadioGroup
                        ],
                        exports: [
                            Radio,
                            RadioGroup
                        ],
                        imports: [
                            common.CommonModule,
                            forms.FormsModule
                        ]
                    },] }
        ];
        return RadioModule;
    }());

    /**
     * Generated bundle index. Do not edit.
     */

    exports.RadioChange = RadioChange;
    exports.RadioGroup = RadioGroup;
    exports.Radio = Radio;
    exports.RadioModule = RadioModule;

    Object.defineProperty(exports, '__esModule', { value: true });

})));
