// carbon-components-angular/bundles/carbon-components-angular-forms.umd.js
/**
 *
 * carbon-angular v0.0.0 | carbon-components-angular-forms.umd.js
 *
 * Copyright 2014, 2021 IBM
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0

 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */


(function (global, factory) {
    typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports, require('@angular/core'), require('@angular/forms'), require('@angular/common'), require('carbon-components-angular/checkbox'), require('carbon-components-angular/toggle'), require('carbon-components-angular/radio'), require('carbon-components-angular/input'), require('carbon-components-angular/button')) :
    typeof define === 'function' && define.amd ? define('carbon-components-angular/forms', ['exports', '@angular/core', '@angular/forms', '@angular/common', 'carbon-components-angular/checkbox', 'carbon-components-angular/toggle', 'carbon-components-angular/radio', 'carbon-components-angular/input', 'carbon-components-angular/button'], factory) :
    (factory((global['carbon-components-angular'] = global['carbon-components-angular'] || {}, global['carbon-components-angular'].forms = {}),global.ng.core,global.ng.forms,global.ng.common,global['carbon-components-angular'].checkbox,global['carbon-components-angular'].toggle,global['carbon-components-angular'].radio,global['carbon-components-angular'].input,global['carbon-components-angular'].button));
}(this, (function (exports,core,forms,common,checkbox,toggle,radio,input,button) { 'use strict';

    // modules
    var NFormsModule = /** @class */ (function () {
        function NFormsModule() {
        }
        NFormsModule.decorators = [
            { type: core.NgModule, args: [{
                        exports: [
                            checkbox.CheckboxModule,
                            toggle.ToggleModule,
                            radio.RadioModule,
                            input.InputModule,
                            button.ButtonModule
                        ],
                        imports: [
                            common.CommonModule,
                            forms.FormsModule,
                            checkbox.CheckboxModule,
                            toggle.ToggleModule,
                            radio.RadioModule,
                            input.InputModule,
                            button.ButtonModule
                        ]
                    },] }
        ];
        return NFormsModule;
    }());

    // export the forms module

    /**
     * Generated bundle index. Do not edit.
     */

    exports.CheckboxModule = checkbox.CheckboxModule;
    exports.Checkbox = checkbox.Checkbox;
    exports.CheckboxChange = checkbox.CheckboxChange;
    exports.CheckboxState = checkbox.CheckboxState;
    exports.ToggleModule = toggle.ToggleModule;
    exports.Toggle = toggle.Toggle;
    exports.ToggleChange = toggle.ToggleChange;
    exports.ToggleState = toggle.ToggleState;
    exports.RadioModule = radio.RadioModule;
    exports.Radio = radio.Radio;
    exports.RadioChange = radio.RadioChange;
    exports.RadioGroup = radio.RadioGroup;
    exports.InputModule = input.InputModule;
    exports.Label = input.Label;
    exports.TextArea = input.TextArea;
    exports.TextInput = input.TextInput;
    exports.ButtonModule = button.ButtonModule;
    exports.Button = button.Button;
    exports.NFormsModule = NFormsModule;

    Object.defineProperty(exports, '__esModule', { value: true });

})));
